src/crimson/osd/shard_services.h

```diff
@@ -188,14 +188,17 @@ class PerShardState {
     return ceph::mono_clock::now() - startup_time;
   }
 
+  const OSDSuperblock& superblock;
```

Comment:
This may be const, but it would be undefined behavior for a shard to read it concurrently with another shard updating it.  You'll need to change the superblock update procedure to be
1) modify OSDSingletonState instance
2) copy new superblock to the PerShardState for each shard.

You'll need an (authoritative, mutable) copy on OSDSingletonState with another (readonly) copy on the PerShardState of each shard.

Reply:
Or, I suppose you only actually need each shard to have the cluster_trim_lower_bound.

Reply:
> Or, I suppose you only actually need each shard to have the cluster_trim_lower_bound.

I prefer the prior. I believe it will be useful in the future as well (besides cluster_trim_lower_bound).